### PR TITLE
[caffe2] Export operators to c10 without including ATen/Tensor.h

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #67096
* #67094

With this change, the only caffe2 files that depend on `ATen/Tensor.h`
are ones that directly use the ATen API. Specifically,
```
[
  "caffe2/CMakeFiles/torch_cuda_cpp.dir/contrib/aten/aten_op_gpu.cc.o",
  "caffe2/CMakeFiles/torch_cpu.dir/core/tensor.cc.o",
  "caffe2/CMakeFiles/torch_cuda_cpp.dir/operators/layer_norm_op.cu.o",
  "caffe2/CMakeFiles/torch_cpu.dir/core/IValueInterface.cc.o",
  "caffe2/CMakeFiles/cuda_tensor_interop_test.dir/__/aten/src/ATen/test/cuda_tensor_interop_test.cpp.o",
  "caffe2/CMakeFiles/torch_cpu.dir/contrib/aten/aten_op.cc.o",
  "caffe2/CMakeFiles/caffe2_pybind11_state_gpu.dir/python/pybind_state.cc.o",
  "caffe2/CMakeFiles/torch_cpu.dir/operators/layer_norm_op.cc.o",
  "caffe2/CMakeFiles/torch_cpu.dir/core/export_c10_op_to_caffe2.cc.o",
  "caffe2/CMakeFiles/torch_cpu.dir/core/export_caffe2_op_to_c10.cc.o",
  "caffe2/CMakeFiles/torch_cpu.dir/operators/enforce_finite_op.cc.o",
  "caffe2/CMakeFiles/torch_cpu.dir/core/operator.cc.o",
  "caffe2/CMakeFiles/tensor_interop_test.dir/__/aten/src/ATen/test/tensor_interop_test.cpp.o",
  "caffe2/CMakeFiles/caffe2_pybind11_state.dir/python/pybind_state.cc.o"
]
```

Differential Revision: [D32289810](https://our.internmc.facebook.com/intern/diff/D32289810)

### CODE REVIEW

**Comment by dagitses** [comment]
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32289810).

**Review by dagitses — APPROVED** [review]

**Comment by dagitses** [comment]
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32289810).

**Comment by dagitses** [comment]
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32289810).

File: caffe2/core/export_caffe2_op_to_c10.h [hunk 1]
```diff
@@ -36,117 +36,31 @@ inline std::vector<caffe2::Tensor> _call_caffe2_op(
...
-      parsed_schema.name(),
-      parsed_schema.overload_name(),
-      std::move(arguments),
-      parsed_schema.returns(),
-      parsed_schema.is_vararg(),
-      parsed_schema.is_varret());
-#endif
-}
+struct TORCH_API RegisterSchema {
+  RegisterSchema(const char *schema_str);
```
Comment (dagitses) [inline caffe2/core/export_caffe2_op_to_c10.h:62]: are you able to mark this explicit?

Comment (peterbell10) [reply caffe2/core/export_caffe2_op_to_c10.h]: done

**Comment by dagitses** [comment]
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32289810).

**Comment by dagitses** [comment]
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32289810).

**Comment by dagitses** [comment]
@dagitses has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32289810).

**Comment by kit1980** [comment]
@peterbell10 @dagitses is this something we want to merge now?

**Comment by peterbell10** [comment]
There is still a caffe2 job in PyTorch CI so I think it still has value.

**Pull request closed without merging**